content/blog/testthat-3-1/index.Rmd

```diff
@@ -0,0 +1,157 @@
+
+-   `expect_snapshot()` and friends gets an experimental new `variant` argument which causes the snapshot to be saved in `_snaps/{variant}/{test}.md` instead of `_snaps/{test}.md`.
+    This allows you to generate (and compare) unique snapshots for different scenarios where the output is otherwise out of your control, like differences across operating systems or R versions.
+
+    ```{r, eval = FALSE}
+    r_version <- function() paste0("R", getRversion()[, 1:2])
+
+    test_that("can capture version nickname", {
+      expect_snapshot(version$nickname, variant = r_version())
+    })
+    ```
+
+Remember that snapshot tests are not run on CRAN by default because they require a human to confirm whether or not a change is a breakage, so you shouldn't rely only on snapshot tests to ensure that your code is correct.
+While it is possible to set `cran = TRUE`, to force snapshot tests to run on CRAN, I don't generally recommend it as snapshots are often vulnerable to minor changes that don't merit breaking your released package.
+
+## Breaking changes
+
+We made one breaking change that affects the [third edition](https://testthat.r-lib.org/articles/third-edition.html).
+Previously, `expect_message()`, `expect_warning()` and `expect_error()` returned the value of the first argument, unless that was `NULL`, when they instead returned the condition object.
+This meant you could write code like this:
+
+```{r, eval = FALSE}
+expect_equal(expect_warning(f(), "warning"), "value")
+```
+
+Now `expect_message()`, `expect_warning()` and `expect_error()` always return the condition object so you need to flip the order of the expectations:
+
+```{r, eval = FALSE}
+expect_warning(expect_equal(f(), "value"), "warning")
+```
+
+This (IMO) is a little easier to read with the pipe:
+
+```{r, eval = FALSE}
+# Or use the pipe:
+f() %>% 
+  expect_equal("value") %>% 
+  expect_warning("warning")
+```
+
+Or with an intermediate object:
+
+```{r, eval = FALSE}
+expect_warning(value <- f(), "warning")
+expect_equal(value, "value")
+```
+
+As with any breaking change, we made this change with great care.
+Fortunately it only affects the 3rd edition, which relatively few packages use, and we submitted PRs to all affected packages on CRAN.
+
+We made this change because it makes testthat more consistent and makes it easier to inspect both the value and the warning.
+This is important because it makes it easier to test functions that produce [custom error objects](https://adv-r.hadley.nz/conditions.html#custom-conditions) that themselves contain meaningful data:
+
+```{r, eval = FALSE}
+library(rlang)
+informative_error <- function() {
+  abort(
+    "An error with extra info",
+    name = "patrice",
+    number = 17
+  )
+}
+
+err <- expect_error(my_function(), class = "package_error_class")
+expect_equal(err$name, "patrice")
+expect_equal(err$number, 17)
+```
+
+Richer conditions are tool that we use within the tidyverse to provide more context in error message.
```

Comment:
```suggestion
Richer conditions are tool that we use within the tidyverse to provide more context in errors, warnings, and messages.
```

Since `expect_error()` always returned the condition and the change only concerns warnings and messages.

Reply:
In addition to lionel's change, the first part should be either "Richer conditions are **a** tool" or "Richer conditions are **tools**"